### ISSUE TITLE
[EPIC] Migrate from libbpf-rs to aya-rs

### ISSUE THREAD

**Issue by vadorovsky** [lockc-project/lockc#49]
Aya allows to write BPF programs in Rust which is natively compiled by Rust compiler.

https://github.com/aya-rs/aya

# TODO

## Upstream

- [x] aya-rs/aya/issues/9
- [x] aya-rs/aya-log/issues/4
- [x] aya-rs/aya/issues/128

## Downstream

First steps, which can be done now:

- [x] #135
- [ ] #136
- [ ] #137

Items blocked by upstream:

- [ ] #138
- [ ] #139

**Comment by vadorovsky** [comment]
Bumping priority of this. Things on aya side are mostly done and we could start an initial rewrite as soon as as any developer has free cycles.

It would be especially nice to get this done before starting writing any new BPF programs.

**Comment by willfindlay** [comment]
I can get started on this as early as this weekend.

**Comment by anakryiko** [comment]
Pardon my curiosity, but would it be ok to ask what exactly is so appealing about writing BPF program source code in Rust compared to C. BPF programs can certainly get quite complicated, but so far I haven't found it restrictive and counter-productive to have to write it in C. C actually lends itself quite nicely to a verifiable, somewhat low-level code that BPF verifier expects.

I've seen lots of production applications using libbpf/libbpf-rs, and Go BPF libraries writing BPF programs using C, and using language of their choice (C++, Rust, Go, what have you) for user-space parts. And I haven't seen much complains about the fact that it's C and not some other language (it's rather compiler doing too clever code transformations and optimizations that's constantly causing troubles, but that's not C fault per se and I'd imagine Rust compiler would try to be as clever as well). On the other hand, all the kernel BPF development and all the new BPF features are always considered along the lines on how they are going to be exposed through C into BPF programs. Same goes for BPF CO-RE and extra C language extensions (mostly through built-ins) are first and foremost designed with the C language in mind. BPF CO-RE is more than field offset relocations, there is an entire family of built-ins that power bpf_core_field_exists(), bpf_core_type_exists(), and a bunch of others (see https://nakryiko.com/posts/bpf-core-reference-guide/#dealing-with-kernel-changes-and-feature-detection for full list). Are those exposed through Rust for BPF program source code?

I'm in no way trying to convince you to stick to C, but I'm just curious what drives people away from the most supported (by kernel BPF developers) and full-featured language frontend (C) into writing that code with Rust, where I suspect a bunch of things might not work or will work not as expected (I'm talking about the BPF program side, mind you)? Honest curiosity, would really appreciate a comprehensive list of reasons. Thank you!

**Comment by vadorovsky** [comment]
Hi @anakryiko. The reasons why I'm considering switching from libbpf to aya are:

* no dependency on LLVM (it can use the internal LLVM of Rust), everything can be done with Cargo
* no need to use bindgen/FFI to have shared structures between userspace and BPF code
* aya only has to use bindgen/FFI to generate bindings to kernel structures, but that's mostly hidden from developers using aya
* I like [aya-log](https://github.com/aya-rs/aya-log) and no necessity for boilerplate code to handle logging with perf event buffer

> Same goes for BPF CO-RE and extra C language extensions (mostly through built-ins) are first and foremost designed with the C language in mind. BPF CO-RE is more than field offset relocations, there is an entire family of built-ins that power bpf_core_field_exists(), bpf_core_type_exists(), and a bunch of others (see https://nakryiko.com/posts/bpf-core-reference-guide/#dealing-with-kernel-changes-and-feature-detection for full list). Are those exposed through Rust for BPF program source code?

`bpf_core_field_exists()` and `bpf_core_type_exists()` type of macros are not currently in aya, as far as I know. There is a general support for CO-RE - that means, all bindings to kernel headers made (automatically) in aya have CO-RE support. I think porting all those macros to aya would be a good idea. Especially because aya already uses `BPF_FIELD_EXISTS` and `BPF_TYPE_EXISTS` in relocation code.

**Comment by vadorovsky** [comment]
So far the only two problems:

* aya-log doesn't work properly. I think the best solution would be doing formatting of arguments in userspace, because doing so in eBPF part annoys the verifier.
* aya doesn't allow to iterate over eBPF maps in the eBPF code

**Comment by vadorovsky** [comment]
# Current status

## Programs to rewrite

- [ ] monitoring new processes - `tp_bpf/sched_process_fork`, `lsm/task_alloc`
- [ ] kernel log audit - `lsm/syslog`
- [ ] mount audit - `lsm/sb_mount` - depends on #114
- [ ] file open - `lsm/file_open` also depends on #114 
- [ ] registering containers
  - [ ] `uprobe/add_container`
  - [ ] `uprobe/delete_containe`
  - [ ] `uprobe/add_process`

## Changes in aya

- [ ] Support for LSM programs - https://github.com/aya-rs/aya/pull/68
- [ ] Support for fentry/fexit programs - https://github.com/aya-rs/aya/pull/139
- Improvements for uprobe programs
  - [ ] `bpf_probe_write_user` function which allows to write data to userspace pointers - https://github.com/aya-rs/aya/pull/142
  - [ ] Still need to test if I can attach to `/proc/self/exe`, but I'm worried I will also have to fix that.
- [ ] Fixing aya-log, do formatting of argumeents in userspace to not annoy BPF verifier. If arguments are just small strings and primitives, I think we can push them to perf buffers as well.

## Other changes I had to make

* I had to split lockc into two binaries, which will be running in the same pod
  * `lockcd-ebpf` - async, based on Tokio, it loads BPF programs, configures and logs them
  * `lockcd-fanofity` - runc observer, it has to be fully synchronous and blocking
    * registration of containers will be done with uprobes - when `lockc-fanotify` will call a function like `create_container`, BPF program in the kernel is going to receive that event and do the registration in BPF maps

To sum it up:
* everything related to BPF - we keep it in tokio
* everything related to monitoring and **freezing** runc - we do it in a blocking binary
